File: worker/src/main/java/com/github/k0kubun/gitstar_ranking/Main.java
package com.github.k0kubun.github_ranking;

import com.github.k0kubun.github_ranking.config.Config;
import com.github.k0kubun.github_ranking.worker.*;
import com.google.common.util.concurrent.ThreadFactoryBuilder;
import io.sentry.Sentry;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;

import javax.sql.DataSource;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Main
{
    private static final Logger LOG = LoggerFactory.getLogger(Main.class);
    private static final int NUM_UPDATE_USER_WORKERS = 2;

    private static final Config config = new Config(System.getenv());

    public static void main(String[] args)
    {
        Sentry.init(System.getenv().get("SENTRY_DSN"));

        ScheduledExecutorService scheduler = buildAndRunScheduler();

        WorkerManager workers = buildWorkers(config);
        workers.start();

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            shutdownAndAwaitTermination(scheduler);
            workers.stop();
        }));
    }

    private static ScheduledExecutorService buildAndRunScheduler()
    {
        ThreadFactory threadFactory = new ThreadFactoryBuilder()
                .setNameFormat("scheduler-%d")
                .setUncaughtExceptionHandler((t, e) -> {
                    Sentry.capture(e);
                    LOG.error("Uncaught exception at scheduler: " + e.getMessage());
                })
                .build();
        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(threadFactory);

        // Schedule at most every 8 hours
        scheduler.scheduleWithFixedDelay(() -> scheduleIfEmpty(config.getQueueConfig().getUserRankingQueue()), 1, 8, TimeUnit.HOURS);
        // scheduler.scheduleWithFixedDelay(() -> { scheduleIfEmpty(config.getQueueConfig().getRepoRankingQueue()); }, 5, 8, TimeUnit.HOURS);
        // scheduler.scheduleWithFixedDelay(() -> { scheduleIfEmpty(config.getQueueConfig().getOrgRankingQueue()); }, 7, 8, TimeUnit.HOURS);

        // Schedule at most every 30 minutes
        scheduler.scheduleWithFixedDelay(() -> scheduleIfEmpty(config.getQueueConfig().getUserStarScanQueue()), 0, 30, TimeUnit.MINUTES);
        scheduler.scheduleWithFixedDelay(() -> scheduleIfEmpty(config.getQueueConfig().getUserFullScanQueue()), 15, 30, TimeUnit.MINUTES);

        return scheduler;
    }

    private static void scheduleIfEmpty(BlockingQueue<Boolean> queue)
    {
        if (queue.size() == 0) {
            try {
                queue.put(true);
            }
            catch (InterruptedException e) {
                Sentry.capture(e);
                LOG.error("Scheduling interrupted: " + e.getMessage());
            }
        }
    }

    private static WorkerManager buildWorkers(Config config)
    {
        DataSource dataSource = config.getDatabaseConfig().getDataSource();

        WorkerManager workers = new WorkerManager();
        for (int i = 0; i < NUM_UPDATE_USER_WORKERS; i++) {
            workers.add(new UpdateUserWorker(dataSource));
        }
        workers.add(new UserRankingWorker(config));
        workers.add(new UserStarScanWorker(config));
        workers.add(new UserFullScanWorker(config));
        return workers;
    }

    // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html
    private static void shutdownAndAwaitTermination(ExecutorService executor)
    {
        executor.shutdown();
        try {
            if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
                executor.shutdownNow();
                if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
                    LOG.error("Failed to shutdown scheduler");
                }
            }
        }
        catch (InterruptedException e) {
            Sentry.capture(e);
            LOG.error("Scheduler shutdown interrupted: " + e.getMessage());
            executor.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
}

File: worker/src/main/java/com/github/k0kubun/gitstar_ranking/Main.kt
package com.github.k0kubun.gitstar_ranking

import com.github.k0kubun.gitstar_ranking.config.Config
import kotlin.jvm.JvmStatic
import io.sentry.Sentry
import java.util.concurrent.ScheduledExecutorService
import com.github.k0kubun.gitstar_ranking.worker.WorkerManager
import java.lang.Thread
import com.google.common.util.concurrent.ThreadFactoryBuilder
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.BlockingQueue
import java.lang.InterruptedException
import com.github.k0kubun.gitstar_ranking.worker.UpdateUserWorker
import com.github.k0kubun.gitstar_ranking.worker.UserRankingWorker
import com.github.k0kubun.gitstar_ranking.worker.UserStarScanWorker
import com.github.k0kubun.gitstar_ranking.worker.UserFullScanWorker
import java.util.concurrent.ExecutorService
import org.slf4j.LoggerFactory

object Main {
    private val LOG = LoggerFactory.getLogger(Main::class.java)
    private const val NUM_UPDATE_USER_WORKERS = 2
    private val config = Config(System.getenv())
    @JvmStatic
    fun main(args: Array<String>) {
        val scheduler = buildAndRunScheduler()
        val workers = buildWorkers(config)
        workers.start()
        Runtime.getRuntime().addShutdownHook(Thread {
            shutdownAndAwaitTermination(scheduler)
            workers.stop()
        })
    }

    private fun buildAndRunScheduler(): ScheduledExecutorService {
        val threadFactory = ThreadFactoryBuilder()
            .setNameFormat("scheduler-%d")
            .setUncaughtExceptionHandler { _: Thread?, e: Throwable ->
                Sentry.capture(e)
                LOG.error("Uncaught exception at scheduler: " + e.message)
            }
            .build()
        val scheduler = Executors.newSingleThreadScheduledExecutor(threadFactory)

        // Schedule at most every 8 hours
        scheduler.scheduleWithFixedDelay({ scheduleIfEmpty(config.queueConfig.userRankingQueue) }, 1, 8, TimeUnit.HOURS)
        // scheduler.scheduleWithFixedDelay(() -> { scheduleIfEmpty(config.getQueueConfig().getRepoRankingQueue()); }, 5, 8, TimeUnit.HOURS);
        // scheduler.scheduleWithFixedDelay(() -> { scheduleIfEmpty(config.getQueueConfig().getOrgRankingQueue()); }, 7, 8, TimeUnit.HOURS);

        // Schedule at most every 30 minutes
        scheduler.scheduleWithFixedDelay({ scheduleIfEmpty(config.queueConfig.userStarScanQueue) }, 0, 30, TimeUnit.MINUTES)
        scheduler.scheduleWithFixedDelay({ scheduleIfEmpty(config.queueConfig.userFullScanQueue) }, 15, 30, TimeUnit.MINUTES)
        return scheduler
    }

    private fun scheduleIfEmpty(queue: BlockingQueue<Boolean>) {
        if (queue.size == 0) {
            try {
                queue.put(true)
            } catch (e: InterruptedException) {
                Sentry.capture(e)
                LOG.error("Scheduling interrupted: " + e.message)
            }
        }
    }

    private fun buildWorkers(config: Config): WorkerManager {
        val dataSource = config.databaseConfig.dataSource
        val workers = WorkerManager()
        for (i in 0 until NUM_UPDATE_USER_WORKERS) {
            workers.add(UpdateUserWorker(dataSource))
        }
        workers.add(UserRankingWorker(config))
        workers.add(UserStarScanWorker(config))
        workers.add(UserFullScanWorker(config))
        return workers
    }

    // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html
    private fun shutdownAndAwaitTermination(executor: ExecutorService) {
        executor.shutdown()
        try {
            if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
                executor.shutdownNow()
                if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
                    LOG.error("Failed to shutdown scheduler")
                }
            }
        } catch (e: InterruptedException) {
            Sentry.capture(e)
            LOG.error("Scheduler shutdown interrupted: " + e.message)
            executor.shutdownNow()
            Thread.currentThread().interrupt()
        }
    }
}
